src/plugins/intel_gpu/include/intel_gpu/primitives/data.hpp

```diff
@@ -4,37 +4,203 @@
+        ib >> do_precision_conversion;
+        if (do_precision_conversion) {
+            ib >> original_size;
+            ib >> make_data(&original_dtype, sizeof(ov::element::Type));
+            ib >> make_data(&curr_dtype, sizeof(ov::element::Type));
+
+            size_t num_dims = 0;
+            ib >> make_data(&num_dims, sizeof(size_t));
+            shape.resize(num_dims);
+            ib >> make_data(shape.data(), num_dims * sizeof(ov::Shape::value_type));
+        } else {
+            original_size = data_size;
+        }
+
+        auto mem_obj = std::make_shared<weights_mem>();
+        mem_obj->shared_buf = std::make_shared<ov::SharedBuffer<std::shared_ptr<ov::MappedMemory>>>(
+            mapped_weights->data() + bin_offset,
+            original_size,
+            mapped_weights);
+
+        if (should_run_transformations()) {
+            run_transformations(mem_obj);
+        }
+        return mem_obj;
+    }
+
+private:
+    bool do_weightless_caching = false;
+    bool do_precision_conversion = false;
+
+    size_t bin_offset = SIZE_MAX;
+    size_t original_size = SIZE_MAX;
+    ov::element::Type original_dtype = ov::element::Type_t::undefined;
+    ov::element::Type curr_dtype = ov::element::Type_t::undefined;
+    ov::Shape shape;
+
+    bool should_run_transformations() {
+        return do_precision_conversion;
+    }
+
+    void run_transformations(std::shared_ptr<weights_mem> mem_obj) {
+        auto orig_constant = std::make_shared<ov::op::v0::Constant>(original_dtype,
+                                                                    shape,
+                                                                    mem_obj->shared_buf->get_ptr(),
+                                                                    mem_obj->shared_buf);
+
+        ov::ParameterVector inputParams;
+        ov::ResultVector results;
+        results.push_back(std::make_shared<ov::op::v0::Result>(orig_constant->output(0)));
+        auto model = std::make_shared<ov::Model>(results, inputParams, "aux");
+
+        ov::pass::Manager manager("Plugin:GPU:weightless_cache_transformations");
+
+        if (do_precision_conversion) {
+            precisions_map fp_convert_precision_map = {
+                {original_dtype, curr_dtype}};
+            type_to_fuse_map empty_fuse_map = {};
+            const bool keep_precision_sensitive_in_fp32 = false;
+            const bool convert_input_output_precision = false;
+            const bool store_original_precision_as_rt_attribute = true;
+            manager.register_pass<ov::pass::ConvertPrecision>(fp_convert_precision_map,
+                                                              empty_fuse_map,
+                                                              keep_precision_sensitive_in_fp32,
+                                                              convert_input_output_precision,
+                                                              store_original_precision_as_rt_attribute);
```

Comment:
What is the typical size of such constants? Maybe we should consider some size limit for this conversion or offload reordering to the GPU in case of large constants? Because in case of large shapes we may overutilize the CPU, which could lead to unexpected/undesired behavior when using the GPU plugin

Reply:
It varies widely, e.g. in llama from a few bytes to 0.5 GB. What kind of unexpected behavior do you have in mind? Performance or something else? If it's the latter, nothing unexpected should happen because the same conversion happens anyway in the first run when cache doesn't yet exist. 

This also could majorly limit the size optimization of the feature which is the whole point.

Reply:
My concern is mainly about the device being used for constants conversion. For sure, we can't avoid CPU usage entirely, but it seems that for the data type conversion of large buffers, the GPU looks like a more appropriate device when "GPU" is selected as the inference device. Also, in some cases, this could provide better performance than the CPU. I'm okay with the current solution as an initial version, but we should probably improve this logic in the future

Reply:
I see your point. In that case implementation of ConvertPrecision itself would have to be modified. I can see that right now it uses avx2 vectorization if possible but I agree offloading to GPU has potential to improve it. But yeah, that sounds like a whole other task and a large one - please correct me if I'm wrong but casting logic for all the different data type is only implemented for the CPU, right?

Reply:
Yes, GPU doesn't support all data types, but it supports most of them, which should be sufficient for most cases. I think we can consider this a further improvement